KAFKA-5547: Return TOPIC_AUTHORIZATION_FAILED error if no describe access for topics

omkreddy opened the pull request.

omkreddy committed:
KAFKA-5547: Return TOPIC_AUTHORIZATION_FAILED error if no describe access for topics

omkreddy commented:
@hachikuji Pls review.

hachikuji commented:
@omkreddy Great, was wondering if you would have time to get to this before the release. I may not have time to get to it this week, but I will be sure to review it next week.

omkreddy commented:
@hachikuji   gentle remainder.

hachikuji commented on core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala:
```diff
@@ -79,7 +79,7 @@ abstract class SaslEndToEndAuthorizationTest extends EndToEndAuthorizationTest {
       fail("Expected exception as consumer2 has no access to group")
     } catch {
-      case _: GroupAuthorizationException => //expected
+      case _: TopicAuthorizationException => //expected
```
We should update the failure message above. By the way, I'm curious why we hit the topic exception first.

omkreddy replied:
Previously we used to get UNKNOWN_TOPIC_OR_PARTITION for metadata fetch request and GroupAuthorizationException during Join group request.  

Now, we get TopicAuthorizationException during metadata fetch request and fail immediately.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -786,7 +786,7 @@ public void handle(OffsetCommitResponse commitResponse, RequestFuture<Void> futu
                         return;
                     } else if (error == Errors.UNKNOWN_TOPIC_OR_PARTITION) {
-                        future.raise(new KafkaException("Partition " + tp + " may not exist or user may not have Describe access to topic"));
+                        future.raise(new KafkaException("Partition " + tp + " may not exist"));
```
Do we still need to say that it _may_ not exist?

omkreddy commented:
@hachikuji  just checking if  we are still aiming this before the code freeze

hachikuji reviewed (commented):
Thanks for the patch. Left a couple initial comments. I think we might be missing some of the handlers. AddPartitionsToTxn is one, but I haven't checked whether there are others. I will look in more detail later today.

hachikuji commented:
@omkreddy Yes, I would like to. I was a bit busy last week, but I think we can get it in today or tomorrow.

omkreddy committed:
Address review comments

omkreddy commented:
Ok, I think, I missed handleAddPartitionToTxnRequest and handleTxnOffsetCommitRequest handlers. I update them.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -786,7 +786,7 @@ public void handle(OffsetCommitResponse commitResponse, RequestFuture<Void> futu
                         return;
                     } else if (error == Errors.UNKNOWN_TOPIC_OR_PARTITION) {
-                        future.raise(new KafkaException("Partition " + tp + " may not exist or user may not have Describe access to topic"));
+                        future.raise(new KafkaException("Partition " + tp + " not exists"));
```
nit: "does not exist" is more grammatical. It could also be that the topic doesn't exist, so maybe we should mention that?

omkreddy committed:
Address review comments

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -264,14 +264,13 @@ class KafkaApis(val requestChannel: RequestChannel,
-                s"on partition $topicPartition failing due to user not having DESCRIBE authorization, but returning UNKNOWN_TOPIC_OR_PARTITION")
-          authorizedForDescribe && exists
+
+      val (authorizedForDescribeTopics, unauthorizedForDescribeTopics) = offsetCommitRequest.offsetData.asScala.toMap.partition {
```
It might not be a major concern, but instead of the multiple calls to `partition`, we could just do a single pass over the partitions and add them to mutable collections.

omkreddy commented:
@hachikuji  Updated code to handle couple of more requests

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1934,8 +1955,7 @@ class KafkaApis(val requestChannel: RequestChannel,
       case RResourceType.TOPIC =>
-        // Don't leak topic name unless the user has describe topic permission
-        if (authorize(session, Describe, new Resource(Topic, resource.name)))
+        if (!authorize(session, Describe, new Resource(Topic, resource.name)) || metadataCache.contains(resource.name))
```
Hmm.. As far as I can tell, this method is used only after failing DescribeConfigs or AlterConfigs authorization. With this patch, shouldn't we just blindly return `TOPIC_AUTHORIZATION_FAILED`.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1424,21 +1439,27 @@ class KafkaApis(val requestChannel: RequestChannel,
 
-    val (authorizedForDeleteTopics, unauthorizedForDeleteTopics) = authorizedForDescribeTopics.partition {
+    val (existingAndAuthorizedForDescribeTopics, nonExistingAndAuthorizedForDescribeTopics) = authorizedForDescribeTopics.partition {
+      case (topicPartition, _) =>metadataCache.contains(topicPartition.topic)
```
nit: space after `=>`

hachikuji commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -521,7 +520,7 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
 
-  @Test(expected = classOf[GroupAuthorizationException])
+  @Test(expected = classOf[TopicAuthorizationException])
   def testConsumeWithNoAccess(): Unit = {
```
Maybe we could mention in the name that this test case is using manual assignment?

hachikuji reviewed (commented):
Left a few more comments.

omkreddy committed:
Address review comments

omkreddy committed:
Address review comments

omkreddy commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -264,25 +264,31 @@ class KafkaApis(val requestChannel: RequestChannel,
+          else
+            nonExistingAndAuthorizedForDescribeTopics += (topicPartition -> partitionData)
+        else
+          unauthorizedForDescribeTopics += (topicPartition -> partitionData)
```
@hachikuji  I tried to implement using mutable collections  and single pass over the map.  If this is OK, or Can you suggest any nicer way to achive this?  Once you confirm, i will implement in other places.

omkreddy replied:
@ijuma Can you suggest a nicer approach for above code block?

ijuma commented on docs/upgrade.html:
```diff
@@ -76,6 +76,10 @@
+    <li>When using an Authorizer and a user doesn't have Describe authorization on a topic, the broker
+        will return TOPIC_AUTHORIZATION_FAILED errors to requests irrespective of topic existence on broker.
+        If the user have Describe authorization and the topic doesn't exists, then the UNKNOWN_TOPIC_OR_PARTITION
+        error code will be returned. </li>
```
Are there any compatibility implications for this? Would a KIP be needed?

hachikuji replied:
The main impact is that the client will fail faster in some cases because we will now get an unambiguous error code. I can't think of a situation where you'd depend on the current behavior... Maybe someone might expect the client to retry internally if it is started prior to granting of permissions?

ijuma replied:
Yeah, seems unlikely, just wanted to double-check. This would be a nice complement to the authentication improvements by @rajinisivaram.

ijuma replied:
@rajinisivaram, any concerns?

rajinisivaram replied:
No concerns, sounds like the right behaviour.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -264,25 +264,31 @@ class KafkaApis(val requestChannel: RequestChannel,
-          nonExistingOrUnauthorizedForDescribeTopics.mapValues(_ => Errors.UNKNOWN_TOPIC_OR_PARTITION)
+          unauthorizedForDescribeTopics.mapValues(_ => Errors.TOPIC_AUTHORIZATION_FAILED) ++
+          nonExistingAndAuthorizedForDescribeTopics.mapValues(_ => Errors.UNKNOWN_TOPIC_OR_PARTITION) ++
+          unauthorizedForReadTopics.mapValues(_ => Errors.TOPIC_AUTHORIZATION_FAILED)
```
Do we need two separate collections that both get mapped to the same error?

omkreddy replied:
yeah..we can use a single collection to handle unauthorized topics. updated the PR

omkreddy committed:
code cleanups

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -276,19 +275,18 @@ class KafkaApis(val requestChannel: RequestChannel,
         else
-          unauthorizedForDescribeTopics += (topicPartition -> partitionData)
+          unauthorizedTopics += topicPartition
       }
```
We could maybe extract a helper method since many requests deal with topics.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -264,25 +264,29 @@ class KafkaApis(val requestChannel: RequestChannel,
-      val (authorizedTopics, unauthorizedForReadTopics) = existingAndAuthorizedForDescribeTopics.partition {
-        case (topicPartition, _) => authorize(request.session, Read, new Resource(Topic, topicPartition.topic))
+      var unauthorizedTopics = Set[TopicPartition]()
+      var nonExistingAndAuthorizedForDescribeTopics = Set[TopicPartition]()
```
Maybe we can use a more concise name like `nonExistingTopics`? Similar for the other methods.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -381,21 +385,29 @@ class KafkaApis(val requestChannel: RequestChannel,
-    val (authorizedRequestInfo, unauthorizedForWriteRequestInfo) = existingAndAuthorizedForDescribeTopics.partition {
-      case (tp, _) => authorize(request.session, Write, new Resource(Topic, tp.topic))
+    for ((topicPartition, memoryRecords) <- produceRequest.partitionRecordsOrFail.asScala) {
+      if (authorize(request.session, Describe, new Resource(Topic, topicPartition.topic)))
```
Thinking about ways to condense these checks. At least maybe we could flatten them?
```
if (!authorize(Describe))
  TOPIC_AUTH
else if (!metadata.contains)
  UNKNOWN
else if (!authorize(Write))
  TOPIC_AUTH
else
  GOOD
```

hachikuji replied:
Briefly discussed this with @ijuma offline, but in some of these cases, I'm not sure the `Describe` check is even needed.  Is it useful to be able to distinguish the UNKNOWN_TOPIC case if the user doesn't have `Write` access anyway? Maybe we could simplify this a little:
```
if (!authorize(Write))
  TOPIC_AUTH
else if (!metadata.contains)
  UNKNOWN
else
  GOOD
```
Thoughts?

omkreddy replied:
Agree.  since (Read, Write, Delete, Alter) operations implies describe operation, It makes sense to do single authorize check at the beginning and then do topic existence check.

Updated the PR. Please check.

hachikuji reviewed (commented):
Left a couple suggestions.

omkreddy committed:
Address review commnets

omkreddy committed:
Address review comments

hachikuji reviewed (commented):
Apologies for one more late comment, but I think we also need to update the OffsetFetch handler. The rest looks good to me.

omkreddy commented:
@hachikuji  In existing code, we were not checking topic read permissions while handling OffsetFetch request.  Was this intentional? or Bug?

I have added a check to topic read permission. With this change, some consumer tests will fail-fast.
Please check the latest commit.

Observations:

1. Currently, we are sending UNKNOWN_TOPIC_OR_PARTITION error for partition errors in OffsetFecthResponse. I have added TOPIC_AUTH_ERROR to OffsetFecthResponse's partition errors. Will there be any compatibility issues?

2. For version 1+ requests we are not doing topic existence check. Do we need to add this check?

hachikuji commented:
@omkreddy Hmm.. I assumed that using `Describe` permission was intentional. Was there a specific reason you felt that needed to be `Read` instead?

> Currently, we are sending UNKNOWN_TOPIC_OR_PARTITION error for partition errors in OffsetFecthResponse. I have added TOPIC_AUTH_ERROR to OffsetFecthResponse's partition errors. Will there be any compatibility issues?

No more so than the other changes I guess. If the user didn't have Describe access previously, they would have probably just retried forever. At least now they can get a useful error.

> For version 1+ requests we are not doing topic existence check. Do we need to add this check?

Actually I had forgotten about this. There is an open PR: #2550. I'll try to revive it, but we probably don't need to address it here.

omkreddy commented:
@hachikuji I was thinking offsetFetch requires Read Permission. I may be comparing with regular Fetch request.  confluent authorization docs table shows OffsetFetch requires READ.
https://docs.confluent.io/current/kafka/authorization.html. 

I think, since we are not reading the topic,  offset fetch request does not require READ permission.

hachikuji commented:
@omkreddy OffsetFetch does require `Read` on the group, but not on the topic. I think we should probably leave the topic authorization check at `Describe`.

hachikuji commented:
Oh, sorry. That table does show `Read` on the topic also... That is a bit odd.. Unless we can think of a good reason to elevate the permission, maybe we should stick with `Describe` for compatibility?

omkreddy committed:
update the OffsetFetch handler

omkreddy commented:
@hachikuji  Agree. Reverted the changes. Updated the PR.  Please check.

hachikuji reviewed (approved):
Thanks for the patch. LGTM. I noticed a couple minor optimizations to avoid unnecessary collection copying. I'll submit a follow-up shortly.
